ASFAchievementManager/AchievementHandler.cs

```diff
@@ -20,23 +20,21 @@ public sealed class AchievementHandler : ClientMsgHandler {
 			switch (packetMsg.MsgType) {
 				case EMsg.ClientGetUserStatsResponse:
 					ClientMsgProtobuf<CMsgClientGetUserStatsResponse> getAchievementsResponse = new(packetMsg);
-					Client.PostCallback(new GetAchievementsCallback(packetMsg.TargetJobID, getAchievementsResponse.Body));
-
+					Client.PostCallback(new GetAchievementsCallback(packetMsg.TargetJobID, getAchievementsResponse.Body, (msg) => msg.eresult));
 					break;
 				case EMsg.ClientStoreUserStatsResponse:
 					ClientMsgProtobuf<CMsgClientStoreUserStatsResponse> setAchievementsResponse = new(packetMsg);
-					Client.PostCallback(new SetAchievementsCallback(packetMsg.TargetJobID, setAchievementsResponse.Body));
-
+					Client.PostCallback(new SetAchievementsCallback(packetMsg.TargetJobID, setAchievementsResponse.Body, (msg) => msg.eresult));
 					break;
 			}
 
 		}
 
-		internal sealed class GetAchievementsCallback : CallbackMsg {
-			internal readonly CMsgClientGetUserStatsResponse Response;
+		internal abstract class AchievementsCallBack<T> : CallbackMsg {
+			internal readonly T Response;
 			internal readonly bool Success;
 
-			internal GetAchievementsCallback(JobID jobID, CMsgClientGetUserStatsResponse msg) {
+			internal AchievementsCallBack(JobID jobID, T msg, Func<T, int> eresult, string error) {
```

Comment:
The type should be `Func<T, EResult>` and not `Func<T, int>`, which also removes the necessity for a cast on line 47.

Reply:
Unfortunatly the response types `CMsgClientGetUserStatsResponse` and `CMsgClientStoreUserStatsResponse` have their `eresult` properties set to int's. So the cast would be required at somepoint, doing it once in the base seemed best.

Reply:
I think it is better to do it twice actually.

The abstract class `AchievementsCallback` contains a protobuf response, but it does not care about the contents or how the contents are represented in there. It receives an extractor-function as a constructor parameter for exactly this reason.

Having the cast inside this abstract class makes it actually care about how the eresult is represented in the protobuf response (e.g. as an integer). After all: the eresult could also be represented as a string (or any other type for that matter). What do you expect the `eresultGetter` to return if one of the responses has it as an integer and one as a string? Let the class that is actually concerned with the content handle the content representation and use the abstraction that the enum type gives you.

Reply:
Oh good point. That actually makes a lot of sense, a second pair of eyes is always helpful! 😄